internal/pkg/gateway/registry.go

```diff
@@ -264,63 +266,54 @@ func (reg *registry) orderers(channel string) ([]*orderer, error) {
 	if err != nil {
 		return nil, err
 	}
-	var orderers []*orderer
 
-	// Get the config
-	config, err := reg.discovery.Config(channel)
-	if err != nil {
-		return nil, err
-	}
-
-	reg.configLock.RLock()
-	defer reg.configLock.RUnlock()
-
-	for _, eps := range config.GetOrderers() {
-		for _, ep := range eps.Endpoint {
-			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
-			if ord, ok := reg.broadcastClients[url]; ok {
-				orderers = append(orderers, ord)
-			}
-		}
+	var orderers []*orderer
+	for _, url := range reg.channelOrderers[channel] {
+		orderers = append(orderers, reg.broadcastClients[url])
 	}
 
 	return orderers, nil
 }
 
 func (reg *registry) registerChannel(channel string) error {
-	// todo need to handle membership updates
 	reg.configLock.Lock() // take a write lock to populate the registry maps
 	defer reg.configLock.Unlock()
 
-	if reg.channelsInitialized[channel] {
-		return nil
-	}
-	// get config and peer discovery info for this channel
-
-	// Get the config
-	config, err := reg.discovery.Config(channel)
-	if err != nil {
-		return err
-	}
-	// get the tlscerts
-	for msp, info := range config.GetMsps() {
-		reg.tlsRootCerts[msp] = info.GetTlsRootCerts()
-	}
+	if valid, exists := reg.configValid.Load(channel); !exists || !valid.(bool) {
+		// Get the config
+		config, err := reg.discovery.Config(channel)
+		if err != nil {
+			return err
+		}
+		// get the tlscerts
+		for msp, info := range config.GetMsps() {
+			reg.tlsRootCerts[msp] = info.GetTlsRootCerts()
+		}
 
-	for mspid, eps := range config.GetOrderers() {
-		for _, ep := range eps.Endpoint {
-			address := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
-			if _, ok := reg.broadcastClients[address]; !ok {
-				// this orderer is new - connect to it and add to the broadcastClients registry
-				tlsRootCerts := reg.tlsRootCerts[mspid]
-				orderer, err := reg.endpointFactory.newOrderer(address, mspid, tlsRootCerts)
-				if err != nil {
-					return err
+		// Get the current set of orderers.  Use existing connections, if available.  Some may have been added, some removed.
+		var channelOrderers []string
+		for mspid, eps := range config.GetOrderers() {
+			for _, ep := range eps.Endpoint {
+				address := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
+				if _, ok := reg.broadcastClients[address]; !ok {
+					// this orderer is new - connect to it and add to the broadcastClients registry
+					tlsRootCerts := reg.tlsRootCerts[mspid]
+					newOrd, err := reg.endpointFactory.newOrderer(address, mspid, tlsRootCerts)
+					if err != nil {
+						return err
+					}
+					reg.broadcastClients[address] = newOrd
+					reg.logger.Infof("Added orderer to registry: %s", address)
 				}
-				reg.broadcastClients[address] = orderer
-				reg.logger.Infof("Added orderer to registry: %s", address)
+				channelOrderers = append(channelOrderers, address)
 			}
 		}
+		reg.channelOrderers[channel] = channelOrderers
+		reg.configValid.Store(channel, true)
```

Comment:
Suppose you have 2 config blocks processed serially, or in a very small amount of time.

The first config block comes in, makes **configValid** be false. 
Further suppose that you get an invocation of **registerChannel** in parallel, and that invocation reaches line 290, so it has read the first config.
Then, suppose the **registerChannel** invocation is delayed, and the second block commits and the callback makes **configValid** be false again. 
Then, when the **registerChannel** reaches line 312, it will set **configValid** for that channel to true, but the cached config will be stale.

Reply:
What I suggest is that you decouple the config parsing and the connection creation (network I/O):

1. Upon a config change, parse the config in the callback and put in the sync.Map the orderer addresses. 
2. Upon a submit (not register), take the latest config, and do the bookkeeping of the connections, while the connections are created asynchronously, and this would let you create them and close them in an almost non-blocking manner.

Reply:
Thanks for this.  I've reimplemented it along the lines you suggest.